helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -122,9 +210,91 @@ public void close() {
     }
+    _baseDataAccessor.unsubscribeDataChanges(_lockPath, this);
     _baseDataAccessor.close();
   }
 
+  @Override
+  public void handleDataChange(String dataPath, Object data) throws Exception {
+    Stat stat = new Stat();
+    ZNRecord readData =
+        _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    LockInfo lockInfo = new LockInfo(readData);
+    // We are the current owner
+    if (lockInfo.getOwner().equals(_userId)) {
+      if (lockInfo.getRequestorId().equals("NONE")|| lockInfo.getPriority() > lockInfo
+          .getRequestorPriority()) {
+        LOG.debug("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
+        // read the lock information again to avoid stale data
+        readData = _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+        // If we are still the lock owner, clean the lock owner field.
+        if (lockInfo.getOwner().equals(_userId)) {
+          ZNRecord znRecord = new ZNRecord(readData);
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.OWNER.name(),
+              LockConstants.DEFAULT_OWNER_TEXT);
+          znRecord.setSimpleField(LockInfo.LockInfoAttribute.MESSAGE.name(),
+              LockConstants.DEFAULT_MESSAGE_TEXT);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.TIMEOUT.name(),
+              LockConstants.DEFAULT_TIMEOUT_LONG);
+          znRecord.setIntField(LockInfo.LockInfoAttribute.PRIORITY.name(),
+              LockConstants.DEFAULT_PRIORITY_INT);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.WAITING_TIMEOUT.name(),
+              LockConstants.DEFAULT_WAITING_TIMEOUT_LONG);
+          znRecord.setLongField(LockInfo.LockInfoAttribute.CLEANUP_TIMEOUT.name(),
+              LockConstants.DEFAULT_CLEANUP_TIMEOUT_LONG);
+          _baseDataAccessor
+              .update(_lockPath, new ZNRecordUpdater(znRecord), AccessOption.PERSISTENT);
+        } else {
+          LOG.info("We are not current lock owner");
+        }
+      }
+    } // We are the current requestor
+    else if (lockInfo.getRequestorId().equals(_userId)) {
+      // In case the owner field is empty, it means previous owner has finished cleanup work.
+      if (lockInfo.getOwner().equals("NONE")) {
+        ZNRecord znRecord = composeNewOwnerRecord();
+        LockUpdater updater = new LockUpdater(new LockInfo(znRecord));
+        _baseDataAccessor.update(_lockPath, updater, AccessOption.PERSISTENT);
+        onAcquiredLockNotification();
+      } else {
+        LOG.info("We do not need to handle this data change");
+      }
+    } // If we are waiting for the lock, but find we are not the requestor any more, meaning we
+    // are preempted by an even higher priority request
+    else if (!lockInfo.getRequestorId().equals("NONE") && !lockInfo.getRequestorId().equals(_userId)
+        && _isPending) {
```

Comment:
Do we need to check this _isPending? I think we can notify anyway if it is pending or not pending.

Reply:
Actually we need, because this callback can also be received by previous owner (who already cleaned up the ownership), and it should not act since it's not pending.